### PR TITLE
Report error when trying to configure a non existing method on MockObject

### PR DESCRIPTION
**Pull request by VincentLanglet**

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Just pushed some commits here.

**Comment by VincentLanglet** [comment]
Thanks @ondrejmirtes. The PR is RTM now.

I don't like the `getClassName` checks, `0` offset and `array_filter` manipulation I had to done.
If you have a better way, I take it !

**Comment by VincentLanglet** [comment]
Test are fine now. :)

File: src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+	{
+		$type = $scope->getType($methodCall->var);
+		if (!($type instanceof IntersectionType)) {
+			return new GenericObjectType(InvocationMocker::class, []);
+		}
+
+		$mockClass = array_filter($type->getTypes(), function (Type $type): bool {
+			return !($type instanceof ObjectType && $type->getClassName() === MockObject::class);
+		});
```
Comment (ondrejmirtes) [inline src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php:37]: Make sure the array isn't empty; please name the variable as a plural.

Comment (ondrejmirtes) [reply src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php:37]: Can you change the condition into:

```php
!$type instanceof TypeWithClassName || $type->getClassName() !== MockObject::class
```

**Comment by ondrejmirtes** [comment]
Oops - one more thing - the new rule isn't registered in rules.neon.

File: src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php:37]: > Make sure the array isn't empty

why ?

I'll return 
```
return new GenericObjectType(InvocationMocker::class, []);
```

This sounds ok to me

Comment (ondrejmirtes) [reply src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php:37]: That's invalid.

**Comment by VincentLanglet** [comment]
This should be ok now @ondrejmirtes :)

**Comment by ondrejmirtes** [comment]
Thank you!

File: src/Rules/PHPUnit/MockMethodCallRule.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace PHPStan\Rules\PHPUnit;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\ObjectType;
+use PHPUnit\Framework\MockObject\InvocationMocker;
```
Comment (Tobion) [inline src/Rules/PHPUnit/MockMethodCallRule.php:11]: This class does not exist and might be causing https://github.com/phpstan/phpstan-phpunit/issues/73

Comment (VincentLanglet) [reply src/Rules/PHPUnit/MockMethodCallRule.php:11]: Should be fixed in https://github.com/phpstan/phpstan-phpunit/pull/74

Tests was running on phpunit ^7.0. The class was existing.

**Pull request merged**